src/helpers.c

```diff
@@ -149,130 +149,108 @@ void concat_str_array(char **output, char *const argv[])
-	pid_t pid;
-	int status = -1;
-
-	pid = fork();
-
-	if (pid == 0) { /* child */
-		if (newstdin >= 0) {
-			if (dup2(newstdin, STDIN_FILENO) == -1) {
-				LOG(NULL, "Could not redirect stdin", "");
-				assert(0);
-			}
-			close(newstdin);
-		}
-		if (newstdout >= 0) {
-			if (dup2(newstdout, STDOUT_FILENO) == -1) {
-				LOG(NULL, "Could not redirect stdout", "");
-				assert(0);
-			}
-			close(newstdout);
-		}
-		if (newstderr >= 0) {
-			if (dup2(newstderr, STDERR_FILENO) == -1) {
-				LOG(NULL, "Could not redirect stderr", "");
-				assert(0);
-			}
-			close(newstderr);
-		}
+	pid_t cmdpid = system_argv_fd_nowait(newstdinfd, newstdoutfd, newstderrfd, cmd);
+	return wait_process_terminate(cmdpid);
+}
 
-		execvp(*argv, argv);
-		LOG(NULL, "This line must not be reached", "");
+pid_t system_argv_fd_nowait(int newstdinfd, int newstdoutfd, int newstderrfd, char *const cmd[])
+{
+	pid_t cmdpid = fork();
+	if (cmdpid == -1) {
+		LOG(NULL, "Failed to fork to execute command", "%s", cmd[0]);
 		assert(0);
-	} else if (pid < 0) {
-		LOG(NULL, "Failed to fork a child process", "");
+	} else if (cmdpid == 0) {
+		set_std_fd(newstdinfd, newstdoutfd, newstderrfd);
+		execvp(*cmd, cmd);
+		LOG(NULL, "Command", "%s failed", cmd[0]);
 		assert(0);
-	} else {
-		pid_t ws = waitpid(pid, &child_exit_status, 0);
-
-		if (ws == -1) {
-			LOG(NULL, "Failed to wait for child process", "");
-			assert(0);
-		}
-
-		if (WIFEXITED(child_exit_status)) {
-			status = WEXITSTATUS(child_exit_status);
-		} else {
-			LOG(NULL, "Child process didn't exit", "");
-			assert(0);
-		}
-
-		if (status != 0) {
-			char *cmdline = NULL;
-
-			concat_str_array(&cmdline, argv);
-			LOG(NULL, "Failed to run command:", "%s", cmdline);
-			free(cmdline);
-		}
 	}
+	return cmdpid;
+}
 
-	return status;
+void set_std_fd(int newstdinfd, int newstdoutfd, int newstderrfd)
+{
+	if (newstdinfd != -1) {
+		move_fd(newstdinfd, STDIN_FILENO);
+	}
+	if (newstdoutfd != -1) {
+		move_fd(newstdoutfd, STDOUT_FILENO);
+	}
+	if (newstderrfd != -1) {
+		move_fd(newstderrfd, STDERR_FILENO);
+	}
 }
 
-int system_argv_pipe(char *const argvp1[], int stdinp1, int stderrp1,
-		     char *const argvp2[], int stdoutp2, int stderrp2)
+void move_fd(int oldfd, int newfd)
 {
-	int statusp2;
-	int pipefd[2];
+	if (dup2(oldfd, newfd) == -1) {
+		LOG(NULL, "Could not create duplicate file descriptor", "%d from %d", newfd, oldfd);
+		assert(0);
+	}
+	if (close(oldfd) == -1) {
+		LOG(NULL, "Could not close file descriptor", "%d", oldfd);
+		assert(0);
+	}
+}
 
-	if (pipe(pipefd)) {
-		LOG(NULL, "Failed to create a pipe", "");
-		return -1;
+int wait_process_terminate(pid_t pid)
+{
+	int status;
+	do {
+		if (waitpid(pid, &status, 0) == -1) {
+			LOG(NULL, "Failed to wait for PID", "%d", pid);
+			return -1;
+		}
+	} while (!WIFEXITED(status) && !WIFSIGNALED(status));
+	// Exit statuses fall in the range of [0, 255].  Make signal statuses fall in a non-overlapping range starting with 256 (128 + (128 + N)).
```

Comment:
Comment doesn't match code. WTERMSIG returns values in [0, 127], so you can't tell if 130 mean exit(130) or signal 2.

Reply:
I misread the documentation here.  This error handling method would only work to properly spread out fatal error signal N, which is coded 128 + N.  It would not work properly for all error signals in the range [0, 255]: http://tldp.org/LDP/abs/html/exitcodes.html